source/validate_atomics.cpp

```diff
@@ -0,0 +1,260 @@
+        if (!_.IsIntScalarType(memory_semantics_type2) ||
+            _.GetBitWidth(memory_semantics_type2) != 32) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+              << spvOpcodeString(opcode)
+              << ": expected Memory Semantics to be 32-bit int";
+        }
+      }
+
+      if (opcode != SpvOpAtomicLoad && opcode != SpvOpAtomicIIncrement &&
+          opcode != SpvOpAtomicIDecrement) {
+        const uint32_t value_type = _.GetOperandTypeId(inst, operand_index++);
+        if (value_type != result_type) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << spvOpcodeString(opcode)
+                 << ": expected Value to be of type Result Type";
+        }
+      }
+
+      if (opcode == SpvOpAtomicCompareExchange ||
+          opcode == SpvOpAtomicCompareExchangeWeak) {
+        const uint32_t comparator_type = _.GetOperandTypeId(inst, operand_index++);
+        if (comparator_type != result_type) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << spvOpcodeString(opcode)
+                 << ": expected Comparator to be of type Result Type";
+        }
+      }
+
+      break;
+    }
+
+    case SpvOpAtomicStore: {
+      const uint32_t pointer_type = _.GetOperandTypeId(inst, 0);
+      uint32_t data_type = 0;
+      uint32_t storage_class = 0;
+      if (!_.GetPointerTypeInfo(pointer_type, &data_type, &storage_class)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Pointer to be of type OpTypePointer";
+      }
+
+      if (!_.IsFloatScalarType(data_type) &&
+          !_.IsIntScalarType(data_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Pointer to be a pointer to int or float "
+               << "scalar type";
+      }
+
+      const uint32_t scope_type = _.GetOperandTypeId(inst, 1);
+      if (!_.IsIntScalarType(scope_type) || _.GetBitWidth(scope_type) != 32) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Scope to be 32-bit int";
+      }
+
+      const uint32_t memory_semantics_type = _.GetOperandTypeId(inst, 2);
+      if (!_.IsIntScalarType(memory_semantics_type) ||
+          _.GetBitWidth(memory_semantics_type) != 32) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Memory Semantics to be 32-bit int";
+      }
+
+      if (opcode == SpvOpAtomicStore) {
+        const uint32_t value_type = _.GetOperandTypeId(inst, 3);
+        if (value_type != data_type) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+              << spvOpcodeString(opcode)
+              << ": expected Value type and the type pointed to by Pointer to"
+              << " be the same";
+        }
+      }
+
+      break;
+    }
+
+    case SpvOpAtomicFlagTestAndSet: {
```

Comment:
I feel this case can be merged with the main case for `SpvOpAtomicLoad` and others to share some checking logic?